connect/src/main/java/com/stripe/android/connect/webview/StripeDownloadListener.kt

```diff
@@ -0,0 +1,97 @@
+package com.stripe.android.connect.webview
+
+import android.app.DownloadManager
+import android.app.DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED
+import android.app.DownloadManager.STATUS_PAUSED
+import android.app.DownloadManager.STATUS_PENDING
+import android.app.DownloadManager.STATUS_RUNNING
+import android.content.Context
+import android.net.Uri
+import android.os.Environment.DIRECTORY_DOWNLOADS
+import android.webkit.DownloadListener
+import android.webkit.URLUtil
+import android.widget.Toast
+import com.stripe.android.connect.R
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+internal class StripeDownloadListener(
+    private val context: Context,
+    private val downloadManager: DownloadManager? =
+        context.getSystemService(Context.DOWNLOAD_SERVICE) as? DownloadManager,
+    private val ioScope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+    // methods below exposed for mocking in tests
+    private val getDownloadManagerRequest: (Uri) -> DownloadManager.Request = { uri ->
```

Comment:
passing in all these lambdas feels a bit strange here. is there some kind of interface you can use here to abstract away the OS level functionality? that way you can just mock the interface.

This will likely make the code a lot easier to follow too

Reply:
Maybe abstracting away and creating components for the following? Then this service could call StripeDownloadManager and you could mock all of these for your tests


StripeDownloadManager:

- startDownload(url, ua, ...,) : DownloadMetadata
- getDownloadProgress(Id): DownloadStatus
- 

DownloadMetadata:
- id
- fileName

ToastManager
- showToast(string)